src/models/release-asset.ts

```diff
@@ -0,0 +1,388 @@
+				options,
+			});
+
+			if (releaseAsset == null) {
+				throw new Error(
+					`Release asset not found '${JSON.stringify(releaseAssetId)}.'`,
+				);
+			}
+
+			return releaseAsset;
+		},
+		async download(id: ReleaseAssetId) {
+			const { asset } = await exports.get(id, { $select: 'asset' });
+			if (asset == null) {
+				throw new Error('Release asset does not contain any uploaded file');
+			}
+
+			return await request.stream({
+				method: 'GET',
+				...parseUrlForRequest(asset.href),
+			});
+		},
+		async upload(
+			uploadParams: WriteReleaseAssetParams,
+			{
+				chunkSize = DEFAULT_MULTIPART_CHUNK_SIZE,
+				parallelUploads = DEFAULT_MULTIPART_PARALLEL_UPLOAD,
+				overwrite = false,
+			}: UploadParams = {},
+		) {
+			const { asset, ...restParams } = uploadParams;
+
+			let size: number;
+			let normalizedParams: Omit<ReleaseAsset['Write'], 'asset'> & {
+				asset: File;
+			};
+
+			if (typeof asset === 'string') {
+				size = await assetHelpers().getFileSize(asset);
+				if (size <= MINIMUM_MULTIPART_SIZE) {
+					const buffer = await assetHelpers().readFileChunk(asset, 0, size);
+					normalizedParams = {
+						...restParams,
+						asset: new File([buffer], getFileBasename(asset), {
+							type: assetHelpers().getMimeType(asset),
+						}),
+					};
+				}
+			} else {
+				size = asset.size;
+				if (size <= MINIMUM_MULTIPART_SIZE) {
+					normalizedParams = uploadParams as Omit<
+						ReleaseAsset['Write'],
+						'asset'
+					> & { asset: File };
+				}
+			}
+
+			// The pattern executed here of first getting the release_asset and then post/patch
+			// is slightly different than most places where we would use upsert, the main reason for that
+			// is to avoid posting a file (higher bandwitch usage) that we know would fail, so we do first a get
+			// and based on that result and the overwrite flag we either patch or post
+			const existingReleaseAsset = await pine.get({
+				resource: 'release_asset',
+				id: { asset_key: restParams.asset_key, release: restParams.release },
+				options: { $select: 'id' },
+			});
+
+			if (existingReleaseAsset != null) {
+				if (overwrite) {
+					if (size <= MINIMUM_MULTIPART_SIZE) {
+						await pine.patch({
+							id: existingReleaseAsset.id,
+							resource: 'release_asset',
+							body: { asset: normalizedParams!.asset },
+						});
+					} else {
+						await uploadMultipartReleaseAsset(existingReleaseAsset.id, asset, {
+							chunkSize,
+							parallelUploads,
+						});
+					}
+					return await exports.get(existingReleaseAsset.id);
+				} else {
+					throw new ReleaseAssetAlreadyExists(
+						restParams.release,
+						restParams.asset_key,
+					);
+				}
+			} else {
+				if (size <= MINIMUM_MULTIPART_SIZE) {
+					return await pine.post({
+						resource: 'release_asset',
+						body: normalizedParams!,
+					});
+				} else {
+					return await createMultipartReleaseAsset(uploadParams, {
+						chunkSize,
+						parallelUploads,
+					});
+				}
+			}
```

Comment:
How about isolating a bit the sections to make it shallower?
The following should be pretty much the exact same amount of lines (ignoring the extra line break) and only repeat the `return await exports.get(existingReleaseAsset.id);` two times.
```suggestion
			if (existingReleaseAsset != null && !overwrite) {
				throw new ReleaseAssetAlreadyExists(
					restParams.release,
					restParams.asset_key,
				);
			}

			if (size <= MINIMUM_MULTIPART_SIZE) {
				// POST body upload upload
				if (existingReleaseAsset != null) {
					await pine.patch({
						id: existingReleaseAsset.id,
						resource: 'release_asset',
						body: { asset: normalizedParams!.asset },
					});
					return await exports.get(existingReleaseAsset.id);
				}
				return await pine.post({
					resource: 'release_asset',
					body: normalizedParams!,
				});
			}

			// Multipart upload
			if (existingReleaseAsset != null) {
				await uploadMultipartReleaseAsset(existingReleaseAsset.id, asset, {
					chunkSize,
					parallelUploads,
				});
				return await exports.get(existingReleaseAsset.id);
			}
			return await createMultipartReleaseAsset(uploadParams, {
				chunkSize,
				parallelUploads,
			});
```